Document how to enable CILogon for a hub (#941)
2i2c-org/infrastructure: docs/howto/configure/auth-management.md

    ```

4. Run the deployer as normal to apply the config.

> # CILogon

Reviewer: ```suggestion
## CILogon
```

> For communities that require authenticating users against their campus identity providers, [CILogon](https://www.cilogon.org) can be used. CILogon authentication is managed through [auth0](https://auth0.com), similar with how the Google and GitHub authentication is setup.

Reviewer: ```suggestion
[CILogon](https://www.cilogon.org) allows us to authenticate users against their campus identity providers. It is managed through [auth0](https://auth0.com), similar to Google and GitHub authentication.
```

> ```{note}

Reviewer: ````suggestion
```{seealso}
See the [CILogon documentation on `Auth0`](https://www.cilogon.org/auth0) for more configuration information.
```
```{note}
````

>   The `USERNAME_KEY` used for the CILogon login is the email, so the JupyterHub username will be set to the email address associated with the CILogon connection and not the CILogon user_id!

Reviewer: ```suggestion
The JupyterHub username will be the email address that users provide when authenticating in CILogon connection. It will not be the CILogon `user_id`! This is because the `USERNAME_KEY` used for the CILogon login is the email address.
```

```

To enable CILogon authentication:

> 1. Explicitly list CILogon as the type of connection we want for a hub, via `auth0.connection`:

Reviewer: ```suggestion
1. List CILogon as the type of connection we want for a hub, via `auth0.connection`:
```

   ```yaml
   auth0:
      connection: CILogon
   ```

> 2. Explicitly list *admin users* for a given hub and only allow the users that match specific identity providers to login into the hub.

Reviewer: ```suggestion
2. Add **admin users** to the hub by explicitly listing their email addresses. Add **allowed users** for the hub by providing a regex pattern that will match to an institutional email address. (see example below)
```

Author: Super nice rephrasing! Thanks @choldgraf :blossom:

  ```{note}

>   Don't forget to allow login to the test user (`deployment-service-check`), otherwise the hub health check performed during deployment will fail.

Reviewer: Isn't this something that we can automatically add at deploy time? It isn't configuration a community would ever want to copy over themselves if they deployed their own hub

Author: Yeah, so if I remember correctly from when I struggled to have a nice way to add the staff by default to the admin&allowed lists, I don't think there's a straightforward way of doing this. And I don't think we should stress out about this as this `username_pattern` thing it's already pretty ugly as it is.

We should be able to get away from this if we manage to automatize the CILogon client creation and move out from Auth0 entirely.

  ```

>   ```yaml
>     config:
>       jupyterhub:
>         hub:
>           config:
>             Authenticator:
>               admin_users:
>                 - user1@campus.edu
>                 - user2@gmail.com
>                 # This will be replaced with the staff's google addresses
>                 - <staff_google_ids>
>               username_pattern: '^(.+@2i2c\.org|.+@campus\.edu|deployment-service-check)$'
>    ```
> 
>   This example does two things:
> 
>   1. Sets the hub admins to be:
> 
>     - the 2i2c staff (identified through their 2i2c email address)
>     - one `@campus.edu` user
>     - one `@gmail.com` user
> 
>   2. Allows logging in only the users that match the `username_pattern` expression:
> 
>     - all `@2i2c.org` email adresses
>     - all `@campus.edu` email addresses
>     - `user2@gmail.com`
>     - the test username, `deployment-service-check`
> 
>   ```{note}
>   All the users listed under `admin_users` need to match the `username_pattern` expression otherwise they won't be allowed to login!
>   ```

Reviewer: ````suggestion

### Example config for CILogon

The following configuration example shows off how to configure hub admins and allowed users:

1. **Hub admins** are these explicit emails:
  - one `@campus.edu` user
  - one `@gmail.com` user
  - the 2i2c staff (identified through their 2i2c email address)

2. **Allowed users** are matched against a pattern, with a few specific addresses added in as well
  - all `@2i2c.org` email adresses
  - all `@campus.edu` email addresses
  - `user2@gmail.com`
  - the test username, `deployment-service-check`

```yaml
config:
  jupyterhub:
    hub:
      config:
        Authenticator:
          admin_users:
            - user1@campus.edu
            - user2@gmail.com
            # This will be replaced with the staff's google addresses
            - <staff_google_ids>
          username_pattern: '^(.+@2i2c\.org|.+@campus\.edu|deployment-service-check)$'
```

```{note}
All the users listed under `admin_users` need to match the `username_pattern` expression otherwise they won't be allowed to login!
```
````

Reviewer: I'm a bit confused where it says that `user2@gmail.com` is also allowed into the hub - I only see it defined in `admin_users`, not in `username_pattern`. Could you clarify?

Author: I meant to add the `user2` in the username pattern, but apparently didn't. I edited you commit suggestion to contain the `user2` example.
